### PR TITLE
net: sockets: ioctl support for FIONBIO and FIONREAD

### PR DESCRIPTION
**Pull request by cfriedt**
In asynchronous server systems, it can be advantageous to be able to look ahead to see how many bytes are available to read on a given socket, without blocking, and without allocating an overly large buffer.

Typically on POSIX systems `ioctl(fd, FIONREAD, &avail)` is used to get the number of bytes available. That method actually predates POSIX though having been a part of BSD systems before that.

This fills one of the upstream gaps needed by the the HTTP GSoC Project.

This change adds
* net: sockets: tcp: split recv_stream into immediate and timed
* add ioctl numbers for `ZFD_IOCTL_FIONBIO`, `ZFD_IOCTL_FIONREAD` to fdtable.h
* posix: ioctl: use fdtable.h `FIONBIO` and `FIONREAD` definitions
* net: sockets: socketpair: support querying bytes available
* tests: net: socketpair: reduce requirements for test
* tests: net: socketpair: add fionread test
* syscalls: include stdarg.h for va_list type
* net: socketpair: support for fionbio
* net: sockets: support fionbio and fionread
* tests: net: socketpair: use a test fixture to deduplicate code
* net: sockets: tcp: test `FIONREAD` with ipv4 and ipv6 sockets
* tests: net: socket: tcp: clear fdtable in after()
* tests: net: socket: udp: clear fdtable in after()
* net: sockets: udp: test `FIONREAD` with ipv4 and ipv6 sockets
* tests: net: socket: tcp: reduce number of file descriptors

There are some other ioctl use cases for networking that may be useful in the future.

### CODE REVIEW

**Comment by cfriedt** [comment]
cc @Emna-Rekik

File: subsys/net/lib/sockets/sockets.c [hunk 1]
```diff
@@ -2530,6 +2581,35 @@ static int sock_ioctl_vmeth(void *obj, unsigned int request, va_list args)
...
+
+		avail = va_arg(args, int *);
+		pkt = k_fifo_peek_head(&ctx->recv_q);
+		if (pkt == NULL) {
+			*avail = 0;
+			return 0;
+		}
+
+		len = net_pkt_get_len(pkt);
+		/* TODO: account for other headers */
```
Comment (jukkar) [inline subsys/net/lib/sockets/sockets.c:2602]: We have net_pkt_ip_hdr_len() helper that should return the IP header length, perhaps we could use it here?

File: tests/net/socket/socketpair/src/_main.c [hunk 2]
```diff
@@ -4,15 +4,42 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#include "_main.h"
+
 #include <zephyr/logging/log.h>
+#include <zephyr/net/socket.h>
 #include <zephyr/ztest.h>
 
-LOG_MODULE_REGISTER(net_test, CONFIG_NET_SOCKETS_LOG_LEVEL);
```
Comment (jukkar) [inline tests/net/socket/socketpair/src/_main.c:10]: Why are we changing the module names?

**Review by jukkar — COMMENTED** [review]
Overall looks good, some minor questions.

File: tests/net/socket/socketpair/src/_main.c (hunk 2 above)
Comment (cfriedt) [reply tests/net/socket/socketpair/src/_main.c:10]: `net_test` was completely non-descriptive of the testsuite. It was likely a typo.

Comment (jukkar) [reply tests/net/socket/socketpair/src/_main.c:10]: There was originally an idea to use net_test as a log module  name for network unit tests, dunno if it still applies. Note that the module name here is only used for logging purposes.

Comment (cfriedt) [reply tests/net/socket/socketpair/src/_main.c:10]: You're right, that's it. I forgot, net subsys logging is actually very well templated.

Comment (cfriedt) [reply tests/net/socket/socketpair/src/_main.c:10]: @jukkar - fixed!

**Review by jukkar — DISMISSED** [review]
LGTM

**Review by nordicjm — DISMISSED** [review]

File: subsys/net/lib/sockets/sockets_tls.c [hunk 3]
```diff
@@ -3215,6 +3215,10 @@ static int tls_sock_ioctl_vmeth(void *obj, unsigned int request, va_list args)
 		return ztls_poll_offload(fds, nfds, timeout);
 	}
 
+	case ZFD_IOCTL_FIONBIO:
+	case ZFD_IOCTL_FIONREAD:
+		return zsock_ioctl(ctx->sock, request, args);
```
Comment (rlubos) [inline subsys/net/lib/sockets/sockets_tls.c:3220]: I don't think that'd be correct for `ZFD_IOCTL_FIONREAD` on TLS socket. The underlying TCP data length may no equal to data length after decryption. We have a dedicated function, `ztls_socket_data_check()`, to check data available after processing by mbed TLS.

**Review by rlubos — COMMENTED** [review]
Looks good, just one concern regarding TLS sock.

File: subsys/net/lib/sockets/sockets_tls.c (hunk 3 above)
Comment (cfriedt) [reply subsys/net/lib/sockets/sockets_tls.c:3220]: Definitely - need to add some test coverage too. I'll see if I can get to this a bit later today.

**Comment by cfriedt** [comment]
Just adding DNM so this doesn't get a fly-by merge until the TLS comment is addressed

File: subsys/net/lib/sockets/sockets_tls.c (hunk 3 above)
Comment (cfriedt) [reply subsys/net/lib/sockets/sockets_tls.c:3220]: @rlubos - I think I'll drop the TLS changes for now. `ztls_socket_data_check()` consistently returns `-EIO` due to `mbedtls_ssl_read()` returning `MBEDTLS_ERR_SSL_BAD_INPUT_DATA` in the case of DTLS.

After looking around, I realized that TLS sockets do not support "peek" operations and would require an intermediate buffer of plain text data in order to determine the size, which is maybe not ideal.

I think that's the case for both TLS and DTLS.

I don't have time right now to look into a possible workaround, but I'll attach the test changes I was planning on adding for TLS to this PR.

**Comment by cfriedt** [comment]
I'll be dropping TLS support for the time being due to implementation detail complexity. I don't have time to find a workaround, and any kind of peek operation apparently requires restructuring `socket_tls.c` to use intermediate buffers, which is maybe not ideal.

net changes
```cpp
diff --git a/subsys/net/lib/sockets/sockets_tls.c b/subsys/net/lib/sockets/sockets_tls.c
index 801a321a20..3af263fbfa 100644
--- a/subsys/net/lib/sockets/sockets_tls.c
+++ b/subsys/net/lib/sockets/sockets_tls.c
@@ -3215,6 +3215,16 @@ static int tls_sock_ioctl_vmeth(void *obj, unsigned int request, va_list args)
                return ztls_poll_offload(fds, nfds, timeout);
        }
 
+       case ZFD_IOCTL_FIONBIO:
+       case ZFD_IOCTL_FIONREAD: {
+               int *avail;
+
+               avail = va_arg(args, int *);
+               *avail = MAX(0, ztls_socket_data_check(ctx));
+
+               return 0;
+       }
+
        default:
                errno = EOPNOTSUPP;
                return -1;
```

tests
```cpp
diff --git a/tests/net/socket/tls/src/main.c b/tests/net/socket/tls/src/main.c
index 0a4dab9f1f..a4e78ffdbd 100644
--- a/tests/net/socket/tls/src/main.c
+++ b/tests/net/socket/tls/src/main.c
@@ -405,6 +405,7 @@ void test_msg_trunc(int sock_c, int sock_s, struct sockaddr *addr_c,
                    socklen_t addrlen_s)
 {
        int rv;
+       int avail;
        uint8_t rx_buf[sizeof(TEST_STR_SMALL) - 1];
        int role = TLS_DTLS_ROLE_SERVER;
        struct test_msg_trunc_data test_data = {
@@ -420,6 +421,11 @@ void test_msg_trunc(int sock_c, int sock_s, struct sockaddr *addr_c,
        rv = bind(sock_s, addr_s, addrlen_s);
        zassert_equal(rv, 0, "server bind failed");
 
+       /* available bytes should be 0 for a newly created socket */
+       avail = 42;
+       zassert_ok(ioctl(sock_s, ZFD_IOCTL_FIONREAD, &avail));
+       zassert_equal(0, avail);
+
        rv = bind(sock_c, addr_c, addrlen_c);
        zassert_equal(rv, 0, "client bind failed");
 
@@ -433,6 +439,12 @@ void test_msg_trunc(int sock_c, int sock_s, struct sockaddr *addr_c,
                              test_msg_trunc_tx_work_handler);
        k_work_reschedule(&test_data.tx_work, K_MSEC(10));
 
+       /* available bytes should be non-zero */
+       k_msleep(100);
+       avail = 0;
+       zassert_ok(ioctl(sock_s, ZFD_IOCTL_FIONREAD, &avail));
+       zassert_true(avail > 0);
+
        memset(rx_buf, 0, sizeof(rx_buf));
        rv = recv(sock_s, rx_buf, 2, ZSOCK_MSG_TRUNC);
        zassert_equal(rv, sizeof(TEST_STR_SMALL) - 1, "MSG_TRUNC flag failed");
@@ -502,6 +514,7 @@ static void test_dtls_sendmsg(int sock_c, int sock_s, struct sockaddr *addr_c,
                              socklen_t addrlen_s)
 {
        int rv;
+       int avail;
        uint8_t rx_buf[sizeof(TEST_STR_SMALL) - 1];
        int role = TLS_DTLS_ROLE_SERVER;
        struct iovec iov[3] = {
@@ -525,6 +538,11 @@ static void test_dtls_sendmsg(int sock_c, int sock_s, struct sockaddr *addr_c,
        rv = bind(sock_s, addr_s, addrlen_s);
        zassert_equal(rv, 0, "server bind failed");
 
+       /* available bytes should be 0 for a newly created socket */
+       avail = 42;
+       zassert_ok(ioctl(sock_s, ZFD_IOCTL_FIONREAD, &avail));
+       zassert_equal(0, avail);
+
        rv = bind(sock_c, addr_c, addrlen_c);
        zassert_equal(rv, 0, "client bind failed");
 
@@ -541,6 +559,12 @@ static void test_dtls_sendmsg(int sock_c, int sock_s, struct sockaddr *addr_c,
 
        k_work_reschedule(&test_data.tx_work, K_MSEC(10));
 
+       /* available bytes should reflect size of datagram */
+       k_msleep(50);
+       avail = 0;
+       zassert_ok(ioctl(sock_s, ZFD_IOCTL_FIONREAD, &avail));
+       zassert_equal(sizeof(TEST_STR_SMALL) - 1, avail);
+
        memset(rx_buf, 0, sizeof(rx_buf));
        rv = recv(sock_s, rx_buf, sizeof(rx_buf), 0);
        zassert_equal(rv, sizeof(TEST_STR_SMALL) - 1, "recv failed");
```

**Comment by cfriedt** [comment]
@rlubos, @jukkar, @nordicjm - please re-ack if you don't mind. I investigated enabling `FIONREAD` support for TLS, but found we would likely need some kind of workaround or to restructure code to use an intermediate buffer, which is likely beyond the scope of what I'd like to do in this PR. Please see my changes above if you would like to investigate further.

**Review by ycsin — DISMISSED** [review]

**Review by nordicjm — DISMISSED** [review]

File: subsys/net/lib/sockets/sockets.c [hunk 4]
```diff
@@ -2530,6 +2581,30 @@ static int sock_ioctl_vmeth(void *obj, unsigned int request, va_list args)
...
+		ssize_t len;
+		struct net_pkt *pkt;
+		struct net_context *ctx = (struct net_context *)obj;
+
+		avail = va_arg(args, int *);
+		pkt = k_fifo_peek_head(&ctx->recv_q);
+		if (pkt == NULL) {
+			*avail = 0;
+			return 0;
+		}
```
Comment (jukkar) [inline subsys/net/lib/sockets/sockets.c:2599]: Should we have a loop here that checks all the pkt in the receive queue, I think we discussed this in networking forum?

File: subsys/net/lib/sockets/sockets_tls.c (hunk 3 above)
Comment (rlubos) [reply subsys/net/lib/sockets/sockets_tls.c]: I don't see anything obvious why would DTLS report an error on `ztls_socket_data_check()`, this needs investigating. I'm generally planning to extend TLS/DTLS tests to cover more socket functionalities, I'm just not there yet. 

In any case, I'm ok with skipping TLS/DTLS sockets implementation as for now.

File: subsys/net/lib/sockets/sockets.c (hunk 4 above)
Comment (cfriedt) [reply subsys/net/lib/sockets/sockets.c:2599]: Yes, thanks for the reminder. I was debugging TLS and forgot to add that testcase.

Comment (cfriedt) [reply subsys/net/lib/sockets/sockets.c]: Loop added! I don't think it's quite perfect yet, but getting there.

**Review by ycsin — DISMISSED** [review]

**Comment by cfriedt** [comment]
Just a minor rebase to pull in the fix in f6fde45e098d6c75d70d0a34cbf15899d6a52903

**Comment by cfriedt** [comment]
@rlubos , @jukkar - I believe this is ready for review now. Just needed to address a corner case.

**Review by rlubos — COMMENTED** [review]
The `recv_stream` rework looks ok, I don't see any obvious problems with the new approach, so as long as all of the tests pass I'm ok with it. It would be good to see some numbers though indeed to see if the rework makes sense in terms of performance.

**Review by jukkar — APPROVED** [review]
LGTM

**Comment by cfriedt** [comment]
@rlubos, @jukkar, @ycsin, @jgl-meta 

Pretty basic benchmark. Time how long it takes to transfer 16kB of lorem_ipsum
https://pastebin.com/pQhcndjY

summary:
* TCP / IP speedup of ~7.5 in terms of time
* far fewer calls to `recv()`
* far fewer cycles spent in `recv()`

before:
```
calls to send(): 427
calls to recv(): 31
EAGAIN in send(): 16
EAGAIN in recv(): 0
cycles in recv(): 114382
end time (ms): 239260
```

after:
```
calls to send(): 459
calls to recv(): 5
EAGAIN in send(): 3
EAGAIN in recv(): 0
cycles in recv(): 96421
end time (ms): 32160
```

**Review by ycsin — APPROVED** [review]

**Review by rlubos — APPROVED** [review]
LGTM, thanks for the numbers!

**Pull request merged**